include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h

```diff
@@ -85,49 +159,67 @@ class sycl_backend
     template <typename NativeUniverseVector>
     sycl_backend(const NativeUniverseVector& v)
     {
+        bool profiling = true;
         global_rank_.reserve(v.size());
         for (auto e : v)
         {
             global_rank_.push_back(e);
+            if (!e.template has_property<sycl::property::queue::enable_profiling>())
+            {
+                profiling = false;
+            }
         }
+        is_profiling_enabled = profiling;
         sgroup_ptr_ = std::make_unique<submission_group>(global_rank_);
     }
 
     template <typename SelectionHandle, typename Function, typename... Args>
     auto
     submit(SelectionHandle s, Function&& f, Args&&... args)
     {
+        constexpr bool report_task_completion = report_info_v<SelectionHandle, execution_info::task_completion_t>;
+        constexpr bool report_task_submission = report_info_v<SelectionHandle, execution_info::task_submission_t>;
+        constexpr bool report_task_time = report_value_v<SelectionHandle, execution_info::task_time_t, report_duration>;
+
         auto q = unwrap(s);
-        if constexpr (report_info_v<SelectionHandle, execution_info::task_submission_t>)
-        {
+
+        if constexpr (report_task_submission)
             report(s, execution_info::task_submission);
-        }
-        if constexpr (report_info_v<SelectionHandle, execution_info::task_completion_t> ||
-                      report_value_v<SelectionHandle, execution_info::task_time_t>)
+
+        if constexpr (report_task_completion || report_task_time)
         {
-            std::chrono::steady_clock::time_point t0;
-            if constexpr (report_value_v<SelectionHandle, execution_info::task_time_t>)
+            const auto t0 = report_clock_type::now();
```

Comment:
I am not sure I agree with the change Sergey proposed to remove `if constexpr` around taking the initial timing snapshot. It makes the code concise, but is that better in this case? Performance-wise it might be OK if we assume the overhead of `std::chrono::steady_clock::now()` is negligible, but it provokes questions like "why are the conditions to set and to use the variable different, including even constexpr checks; is it by purpose?".

Maybe using the ternary operator `?:` is a good compromise here; it is not constexpr but hopefully the compiler can optimize it anyway. But if I am the only one concerned and others are not, then fine to keep the code as now.

Reply:
I think it's ok as is. Let's keep as is. We can always revisit it if we see some performance impact later.